Python/optimizer_analysis.c

```diff
@@ -206,14 +625,103 @@ remove_globals(_PyInterpreterFrame *frame, _PyUOpInstruction *buffer,
                 builtins = func->func_builtins;
                 break;
             }
-            case _JUMP_TO_TOP:
-            case _EXIT_TRACE:
-                return 1;
+            default:
+                if (op_is_end(opcode)) {
+                    return 1;
+                }
+                break;
         }
     }
     return 0;
 }
 
+static int
+uop_redundancy_eliminator(
+    PyCodeObject *co,
+    _PyUOpInstruction *trace,
+    int trace_len,
+    int curr_stacklen
+)
+{
+
+    _Py_UOpsAbstractInterpContext ctx;
+
+    if (abstractcontext_init(
+        &ctx,
+        co, curr_stacklen,
+        trace_len) < 0) {
+        goto error;
+    }
+    _PyUOpInstruction *curr = NULL;
+    _PyUOpInstruction *end = NULL;
+    int status = 0;
+
+    curr = trace;
+    end = trace + trace_len;
+    ;
+    while (curr < end && !op_is_end(curr->opcode)) {
```

Comment:
Could you use a for loop? It is more idiomatic C.
```suggestion
    for (_PyUOpInstruction *curr = trace; curr < trace + trace_len && !op_is_end(curr->opcode); curr++) {
```
If you handle testing for "end" uops in the switch then `for (_PyUOpInstruction *curr = trace; ; curr++) {` will do.

Reply:
There should always be an end uop, so maybe:
```C
for (_PyUOpInstruction *curr = trace; !op_is_end(curr->opcode); curr++) {
    assert(curr < trace + trace_len);
```